### DISCUSSION TITLE
transform jcl into mvscmd

### DISCUSSION THREAD

**Discussion by rpagroot** [IBM/zoau-samples#22] [Q&A] [answered]
Hi,

Tried to transform a piec of jcl into a working mvscmd.execute_authorized command.  The problem is the "PARM='++111.240' " part.

Properly working JCL:

```
//S010     EXEC PGM=ISFAFD,PARM='++111,240'
//ISFOUT   DD DSN=MYHLQ.TMPDS,DISP=(NEW,PASS),
//             SPACE=(CYL,(5,1),RLSE),
//             DCB=(RECFM=FBA,LRECL=244,BLKSIZE=24400)
//ISFIN    DD *
PRE GR*
OWNER GR*
SORT JOBCRDATE D
FIL JOBID *
ST
```

Not yet correctly working execute_authorized  statement:

```
rc = execute_authorized(pgm="ISFAFD", pgm_args="PARM=++111,240", dds=dd_list)
                                                                                                              
```

The output file is filled up with hex 00 chars as off position 132 (default output for SDSF) indicating the "PARM='++111,240'"   is not effective and it ends up in the SYSIN somehow.    

How should I invoke this properly?

Kind regards,

Richard Grooot

**Comment by EnsonoStevenPerva** [comment]
Richard,

Have you tried `pgm_args` without the `PARM=` section, like this?

`rc = execute_authorized(pgm="ISFAFD", pgm_args="++111,240", dds=dd_list)`

Without your full Python listing its hard to test this one.

Steven

**Reply by EnsonoStevenPerva** [reply]
Interestingly, when I invoke ISFAFD via `mvscmd.execute_authorized` method, I get a bizarre response:

`UnicodeDecodeError: 'utf-8' codec can't decode byte 0x86 in position 277: invalid start byte`

But if I gather up all of the information to execute it from the native ZOAU Command mvscmdauth:

`mvscmdauth --args='++111,240' --pgm=ISFAFD  --ISFOUT="*" --ISFIN="/tmp/test.in"`
(making sure `/tmp/test.in` contains your TSFIN data)

My screen is scrolled with SDSF screens, which leads me to believe this is getting closer to functioning. I believe something in the response is causing `call_zoau_library` to have a fit, that might benefit from one of the ZOAU pros to look at. Here's a test harness:

**File test.in:**
```
PRE GR*
OWNER GR*
SORT JOBCRDATE D
FIL JOBID *
ST
```

**File test_harness.py:**
```
from zoautil_py import mvscmd
from zoautil_py.types import DDStatement, DatasetDefinition, FileDefinition

dd_list = []
dd_list.append(DDStatement(name='ISFOUT', definition='*'))
dd_list.append(DDStatement(name='ISFIN', definition=FileDefinition('/tmp/test.in')))

rc = mvscmd.execute_authorized(pgm="ISFAFD", pgm_args="++111,240", dds=dd_list)
print(rc)
```

Hope this helps dig to the root of the problem here.

Steven

**Comment by rpagroot** [comment, marked as answer]
Hi Steven,

This works fine indeed:

`rc = execute_authorized(pgm="ISFAFD", pgm_args="++111,240", dds=dd_list) `

Thanx!   

You've been using a zfs file (presumeably UTF8)  instead of EBCDIC datasets.
This is my ddlist: 

```
sysin = (f"PRE {prefix}\n"       
         f"OWNER {owner}\n"      
         f"SORT JOBCRDATE D\n"
         f"FIL JOBID {jobid}\n"  
         f"ST")                                                                               

z     = create(creafn(), type="SEQ")        
rcw   = write(z.name, sysin, append=False)  
o     = create(creafn(),type='SEQ',record_length=244,block_size=24400,record_format='FBA')   
                                                      
dd_list = list()                                                                    
dd_list.append(DDStatement('ISFOUT',DatasetDefinition(o.name)))
dd_list.append(DDStatement('ISFIN',DatasetDefinition(z.name)))     
```

**Reply by EnsonoStevenPerva** [reply]
Thanks for the follow-up. I'm glad that worked out for you! I learned a good bit from the code you've provided here in creating your datasets, so thanks for sharing that as well.

**Comment by MikeFultonDev** [comment]
Great discussion! Some of the MVS programs are a little harder to interface with - glad you figured out the right incantation for the dataset.
